packages/abi-coder/package.json

```diff
@@ -6,16 +6,18 @@
   "typedoc": {
     "entryPoint": "src/index.ts"
   },
-  "main": "src/index.ts",
+  "exports": {
+    ".": "./src/index.ts"
+  },
   "publishConfig": {
     "main": "dist/index.js",
     "module": "dist/index.mjs",
     "types": "dist/index.d.ts",
-    "typings": "dist/index.d.ts",
     "exports": {
       ".": {
         "require": "./dist/index.js",
-        "default": "./dist/index.mjs"
+        "default": "./dist/index.mjs",
+        "types": "./dist/index.d.ts"
```

Comment:
Great question.

We have two things here:

 1. The `publishConfig` ([docs](https://pnpm.io/package_json#publishconfig)) is a `pnpm` feature _(not an `npm` one)_ that lets us use an optimal configuration for development and override it with a better alternative for and right before publishing.
 1. The `exports` ([docs](https://nodejs.org/api/packages.html#exports)) is a NodeJS feature that makes it possible to have **_multiple entry-points_**, while **_exporting different versions_** of each one _(mainly `cjs`, `esm`, and `iife`)_ for increased compatibility.

In the example below, we point the `exports` field to our **_Typescript_** file _(optimal for development)_, which enables us to have a full-blown cross-package `go to definition` feature working nicely inside VSCode. Without this, we'd be sent to the `*.d.ts` definitions instead of the source `*.ts` ones, breaking the flow.

```ts
"exports": {
  ".": "./src/index.ts"
,
```

Then we override it for publishing with the following:

```ts
"publishConfig": {
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "exports": {
    ".": {
      "require": "./dist/index.js",
      "default": "./dist/index.mjs",
      "types": "./dist/index.d.ts"
    }
  }
}
```

The fields `main`, `module`, and `types` are a fallback for legacy versions of NodeJS that don't understand the `exports` field. Since the `abi-coder` package exports only a _**single entry point**_ (`index.*`), this is fully equivalent.

However, some packages export multiple entry points _(`abi-typegen`, `fuels`, `versions`, and `wallet`)_, something that is fully supported when using the `exports` syntax but falls short when using the legacy properties approach that allows only a **_single entry point_**.

In those cases, we still have the fallback properties in place, but they'll export only the principal entry point, leaving the others out of the equation. Think of this as a type of _Progressive Enhancement_ approach, targeting newer versions with all functionalities while degrading them to what is possible for legacy ones.

Finally,

You will notice that I removed all legacy fields from the `package.json` ROOT _(in favor of the `exports` syntax)_, leaving them only inside the `publishConfig` container. This is to reduce the confusion between both approaches and **_also_** to straighten up some tiny conflicts I found happening when the two are enabled simultaneously _while pointing to `*.ts` files instead of `*.m?js`_ — and since we don't rely on legacy versions for development, we don't need to care about these _[legacy]_ properties while developing.

Reply:
Ok, I misspoke, `publishConfig` is indeed an `npm` thing (_as well_):
 - https://docs.npmjs.com/cli/v7/configuring-npm/package-json#publishconfig